src/emitter/json.rs

```diff
@@ -67,20 +67,22 @@ impl JsonEmitter {
             let mut expected_end_line = expected_begin_line;
             let mut original_line_counter = 0;
             let mut expected_line_counter = 0;
-            let mut original_lines = vec![];
-            let mut expected_lines = vec![];
+            let mut original = String::new();
+            let mut expected = String::new();
 
             for line in mismatch.lines {
                 match line {
                     DiffLine::Expected(msg) => {
                         expected_end_line = expected_begin_line + expected_line_counter;
                         expected_line_counter += 1;
-                        expected_lines.push(msg)
+                        expected.push_str(&msg);
+                        expected.push('\n');
```

Comment:
I remember when we first added the json output that the line numbers were surprisingly tricky to get right, especially when there are changes that include multi-lined snippets collapsed into one, and vice versa, and newlines inserted and/or removed. 

I'd suggest using [this file](https://github.com/rust-lang/rustfmt/blob/master/tests/writemode/source/fn-single-line.rs) as a test (make sure you set `fn_single_line` to true, either in a config file or via `--config fn_single_line=true`) and just checking that the line numbers in the json output match up with what rustfmt actually emits in the standard file emit mode. 

I'm not suggesting there's necessarily an issue nor change needed with the line numbers as part of this (it's entirely possible there's bugs with the current version), but just want to make sure we test with some more realistic types of files and not just single function types of snippets.

Reply:
Here's the output for that file with `--config fn_single_line=true`:

```json
[
  {
    "name": "/Users/timh/d1/fn-single-line.rs",
    "mismatches": [
      {
        "original_begin_line": 5,
        "original_end_line": 7,
        "expected_begin_line": 5,
        "expected_end_line": 5,
        "original": "fn foo_expr() {\n    1\n}\n",
        "expected": "fn foo_expr() { 1 }\n"
      },
      {
        "original_begin_line": 9,
        "original_end_line": 11,
        "expected_begin_line": 7,
        "expected_end_line": 7,
        "original": "fn foo_stmt() {\n    foo();\n}\n",
        "expected": "fn foo_stmt() { foo(); }\n"
      },
      {
        "original_begin_line": 13,
        "original_end_line": 15,
        "expected_begin_line": 9,
        "expected_end_line": 9,
        "original": "fn foo_decl_local()  {\n    let z = 5;\n   }\n",
        "expected": "fn foo_decl_local() { let z = 5; }\n"
      },
      {
        "original_begin_line": 17,
        "original_end_line": 19,
        "expected_begin_line": 11,
        "expected_end_line": 11,
        "original": "fn    foo_decl_item(x: &mut i32) {\n    x = 3;\n}\n",
        "expected": "fn foo_decl_item(x: &mut i32) { x = 3; }\n"
      },
      {
        "original_begin_line": 21,
        "original_end_line": 21,
        "expected_begin_line": 13,
        "expected_end_line": 13,
        "original": "   fn empty()     {\n",
        "expected": "fn empty() {}\n"
      },
      {
        "original_begin_line": 23,
        "original_end_line": 23,
        "expected_begin_line": 15,
        "expected_end_line": 15,
        "original": "}\n",
        "expected": "fn foo_return() -> String { \"yay\" }\n"
      },
      {
        "original_begin_line": 25,
        "original_end_line": 29,
        "expected_begin_line": 17,
        "expected_end_line": 20,
        "original": "fn foo_return() -> String {\n    \"yay\"\n}\n\nfn foo_where() -> T where T: Sync {\n",
        "expected": "fn foo_where() -> T\nwhere\n    T: Sync,\n{\n"
      },
      {
        "original_begin_line": 33,
        "original_end_line": 37,
        "expected_begin_line": 24,
        "expected_end_line": 24,
        "original": "fn fooblock() {\n    {\n        \"inner-block\"\n    }\n}\n",
        "expected": "fn fooblock() { { \"inner-block\" } }\n"
      },
      {
        "original_begin_line": 64,
        "original_end_line": 66,
        "expected_begin_line": 51,
        "expected_end_line": 51,
        "original": "fn lots_of_space                      ()                                                           {\n                           1                 \n}\n",
        "expected": "fn lots_of_space() { 1 }\n"
      },
      {
        "original_begin_line": 71,
        "original_end_line": 72,
        "expected_begin_line": 56,
        "expected_end_line": 56,
        "original": "    fn dummy(&self) {\n    }\n",
        "expected": "    fn dummy(&self) {}\n"
      },
      {
        "original_begin_line": 75,
        "original_end_line": 75,
        "expected_begin_line": 59,
        "expected_end_line": 60,
        "original": "trait CoolerTypes { fn dummy(&self) { \n",
        "expected": "trait CoolerTypes {\n    fn dummy(&self) {}\n"
      },
      {
        "original_begin_line": 77,
        "original_end_line": 77,
        "expected_begin_line": 62,
        "expected_end_line": 62,
        "original": "}\n",
        "expected": ""
      },
      {
        "original_begin_line": 79,
        "original_end_line": 79,
        "expected_begin_line": 63,
        "expected_end_line": 66,
        "original": "fn Foo<T>() where T: Bar {\n",
        "expected": "fn Foo<T>()\nwhere\n    T: Bar,\n{\n"
      }
    ]
  }
]
```

And here's the formatted file:

```
   1  // rustfmt-fn_single_line: true
   2  // rustfmt-emit_mode: checkstyle
   3  // Test single-line functions.
   4  
   5  fn foo_expr() { 1 }
   6  
   7  fn foo_stmt() { foo(); }
   8  
   9  fn foo_decl_local() { let z = 5; }
  10  
  11  fn foo_decl_item(x: &mut i32) { x = 3; }
  12  
  13  fn empty() {}
  14  
  15  fn foo_return() -> String { "yay" }
  16  
  17  fn foo_where() -> T
  18  where
  19      T: Sync,
  20  {
  21      let x = 2;
  22  }
  23  
  24  fn fooblock() { { "inner-block" } }
  25  
  26  fn fooblock2(x: i32) {
  27      let z = match x {
  28          _ => 2,
  29      };
  30  }
  31  
  32  fn comment() {
  33      // this is a test comment
  34      1
  35  }
  36  
  37  fn comment2() {
  38      // multi-line comment
  39      let z = 2;
  40      1
  41  }
  42  
  43  fn only_comment() {
  44      // Keep this here
  45  }
  46  
  47  fn aaaaaaaaaaaaaaaaa_looooooooooooooooooooooong_name() {
  48      let z = "aaaaaaawwwwwwwwwwwwwwwwwwwwwwwwwwww";
  49  }
  50  
  51  fn lots_of_space() { 1 }
  52  
  53  fn mac() -> Vec<i32> { vec![] }
  54  
  55  trait CoolTypes {
  56      fn dummy(&self) {}
  57  }
  58  
  59  trait CoolerTypes {
  60      fn dummy(&self) {}
  61  }
  62  
  63  fn Foo<T>()
  64  where
  65      T: Bar,
  66  {
  67  }
```

It looks correct to me.